src/plugins/spec/spec-new.c

```diff
@@ -0,0 +1,225 @@
+}
+#else
+/**
+ * Check whether the {@link otherKey} matches the {@link specKey}.
+ *
+ * @param specKey specification key
+ * @param otherKey the other key to match the specification key with
+ * @return 1 - if the keys match
+ * 	   0 - if the keys do not match
+ */
+static bool specMatches (Key * specKey, Key * otherKey)
+{
+	// ignore namespaces for globbing
+	Key * globKey = keyNew (strchr (keyName (otherKey), '/'), KEY_END);
+	bool matches = elektraKeyGlob (globKey, strchr (keyName (specKey), '/')) == 0;
+	keyDel (globKey);
+	return matches;
+}
+#endif
+
+/**
+ * Appends a key to the KeySet with the default value specified in the meta key (meta:/default) in the
+ * {@link specKey}.
+ *
+ * The default key is added to the `default:/` namespace.
+ *
+ * @param ks the key store to append the new default key to
+ * @param parentKey of the key to appended
+ * @param specKey specification key with meta data of the new default key
+ */
+static void addDefaultKeyIfNotExists (KeySet * ks, Key * parentKey, Key * specKey)
+{
+	const Key * defaultMetaKey = keyGetMeta (specKey, "default");
+	const char * defaultValue = keyString (defaultMetaKey);
+
+	const char * parentKeyName = keyName (parentKey);
+	const char * specKeyName = keyBaseName (specKey);
+
+	Key * newDefaultKey = keyNew (elektraFormat ("%s:/%s/%s", "default", parentKeyName, specKeyName), KEY_VALUE, defaultValue, KEY_END);
+	keyCopyAllMeta (newDefaultKey, specKey);
+
+	ksAppendKey (ks, newDefaultKey);
+}
+
+static bool isRequired (Key * specKey)
+{
+	const Key * key = keyGetMeta (specKey, "require");
+
+	if (key == 0)
+	{
+		return false;
+	}
+
+	const char * keyValue = keyString (key);
+
+	return elektraStrCmp (keyValue, "true") == 0;
+}
+
+static bool hasDefault (Key * specKey)
+{
+	return keyGetMeta (specKey, "default") != 0;
+}
+
+/**
+ * Extracts all the specification keys from the KeySet. After applying this method the {@link ks}
+ * contains no specification namespace.
+ *
+ * It returns the extracted specification keys.
+ *
+ * @param ks the KeySet to extract specification keys from
+ *
+ * @return the extracted specification keys
+ */
+static KeySet * extractSpecKeys (KeySet * ks)
+{
+	KeySet * specKeys = ksNew (0, KS_END);
+	for (elektraCursor it = 0; it < ksGetSize (ks); it++)
+	{
+		Key * current = ksAtCursor (ks, it);
+		if (keyGetNamespace (current) == KEY_NS_SPEC)
+		{
+			ksAppendKey (specKeys, current);
+		}
+	}
+
+	Key * specKey = keyNew ("spec:/", KEY_END);
+	ksDel (ksCut (ks, specKey));
+	keyDel (specKey);
+
+	return specKeys;
+}
+
+/**
+ * Copies all meta keys from the {@link specKey} to the provided {@link key}.
+ *
+ * @param key the key to copy the meta data too
+ * @param specKey the specification key to copy meta data from
+ *
+ * @return 0 - in case the copying was successful
+ * 	  -1 - if the copying was unsuccessful
+ */
+static int copyMeta (Key * key, Key * specKey)
+{
+	KeySet * metaKeys = ksDup (keyMeta (specKey));
```

Comment:
Why are you duplicating the returned KeySet?

Reply:
You are right, should not be needed here.